File: atcoder/abc/abc317/c/main.py
N, M = map(int, input().split())
G = [[] for _ in range(N)]
for _ in range(M):
  A, B, C = map(int, input().split())
  A-=1
  B-=1
  G[A].append((B, C))
  G[B].append((A, C))

import sys
sys.setrecursionlimit(10**6)
ans = 0
seen = [0]*N
def dfs(x, dist):
  global ans
  seen[x] = 1
  for nx, d in G[x]:
    if not seen[nx]:
      seen[nx] = 1
      dfs(nx, dist+d)
      seen[nx] = 0
  ans = max(ans, dist)
  seen[x] = 0

for i in range(N):
  dfs(i, 0)
print(ans)

File: atcoder/abc/abc317/c/main.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

struct edge{
  ll to;ll d;
};
void show(unordered_set<ll> vis){
  for(auto x: vis) cout << x << ' ';
  cout << endl;
}


void dfs(ll &ans, vector<vector<edge>> &G, unordered_set<ll> visited, vector<bool> &seen, ll x, ll dist){
  ans = max(ans, dist);
  cout << ans << ' ' << x << endl;
  show(visited);
  seen[x] = true;
  for(auto nx: G[x]){
    if(visited.count(nx.to)==0){
      visited.insert(nx.to);
      dist += nx.d;
      dfs(ans, G, visited, seen, x, dist);
      visited.erase(nx.to);
    }
  }
  visited.erase(x);
}

int main(){
  ll N, M;
  cin >> N >> M;
  vector<vector<edge>> G(N, vector<edge>());
  ll A, B, C;
  for(int i=0;i<M;i++){
    cin >> A >> B >> C;
    A--;B--;
    G[A].push_back({B, C});
    G[B].push_back({A, C});
  }
  ll ans = 0;
  vector<bool> seen(N, false);
  for(int i=0;i<N;i++){
    unordered_set<ll> visited;
    visited.insert(i);
    cout << i << endl;
    dfs(ans, G, visited, seen, 0, 0);
  }
  cout << ans << endl;
}